src/mysql_vm_helpers.py

```diff
@@ -280,29 +280,48 @@ def write_mysqld_config(self) -> None:
             content=content_str,
         )
 
-    def setup_logrotate_and_cron(self) -> None:
-        """Create and write the logrotate config file."""
+    def setup_logrotate_and_cron(self, enabled_log_files: Iterable) -> None:
+        """Setup log rotation configuration for text files.
+
+        Args:
+            enabled_log_files: a iterable of enabled text logs
+        """
         logger.debug("Creating logrotate config file")
+        config_path = "/etc/logrotate.d/flush_mysql_logs"
+        script_path = f"{self.charm.charm_dir}/logrotation.sh"
+        cron_path = "/etc/cron.d/flush_mysql_logs"
+        logs_dir = f"{CHARMED_MYSQL_COMMON_DIRECTORY}/var/log/mysql"
 
         with open("templates/logrotate.j2", "r") as file:
             template = jinja2.Template(file.read())
 
-        rendered = template.render(
+        logrotate_conf_content = template.render(
             system_user=MYSQL_SYSTEM_USER,
-            snap_common_directory=CHARMED_MYSQL_COMMON_DIRECTORY,
+            log_dir=logs_dir,
             charm_directory=self.charm.charm_dir,
             unit_name=self.charm.unit.name,
+            enabled_log_files=enabled_log_files,
         )
 
-        with open("/etc/logrotate.d/flush_mysql_logs", "w") as file:
-            file.write(rendered)
+        self.write_content_to_file(
+            config_path, logrotate_conf_content, owner="root", permission=0o644
+        )
 
-        cron = (
-            "* 1-23 * * * root logrotate -f /etc/logrotate.d/flush_mysql_logs\n"
-            "1-59 0 * * * root logrotate -f /etc/logrotate.d/flush_mysql_logs\n"
+        with open("templates/run_log_rotation.sh.j2", "r") as file:
+            template = jinja2.Template(file.read())
+
+        logrotation_script_content = template.render(
+            log_path=f"{CHARMED_MYSQL_COMMON_DIRECTORY}/var/log/mysql",
+            enabled_log_files=enabled_log_files,
+            logrotate_conf=config_path,
+            owner=MYSQL_SYSTEM_USER,
+            group=MYSQL_SYSTEM_USER,
         )
-        with open("/etc/cron.d/flush_mysql_logs", "w") as file:
-            file.write(cron)
+
+        self.write_content_to_file(script_path, logrotation_script_content, permission=0o550)
+
+        cron_content = f"* 1-23 * * * root {script_path}\n" f"1-59 0 * * * root {script_path}\n"
```

Comment:
Nit: could we have these two sub-strings in different lines somehow? I think they would be easier to read.

Reply:
I just now notice it was 2 substr. Since it's pretty small, the formater (ruff) will rewrite as a single liner, and adding a line by, for instance putting in a tuple and using `"\n".join(substring_tuple)`, seems unnecessary this simple case.

Reply:
> Since it's pretty small, the formater (ruff) will rewrite as a single liner

Ugh, annoying. I wonder whether there is a rule to change that.

> Using "\n".join(substring_tuple) seems unnecessary this simple case.

Agreed.